core/src/main/java/ai/timefold/solver/core/impl/domain/variable/declarative/DefaultShadowVariableSessionFactory.java

```diff
@@ -36,10 +48,112 @@ public static <Solution_> VariableReferenceGraph<Solution_> buildGraph(
             SolutionDescriptor<Solution_> solutionDescriptor,
             VariableReferenceGraphBuilder<Solution_> variableReferenceGraphBuilder, Object[] entities,
             IntFunction<TopologicalOrderGraph> graphCreator) {
-        var declarativeShadowVariableDescriptors = solutionDescriptor.getDeclarativeShadowVariableDescriptors();
-        if (declarativeShadowVariableDescriptors.isEmpty()) {
-            return EmptyVariableReferenceGraph.INSTANCE;
+        var graphStructure = GraphStructure.determineGraphStructure(solutionDescriptor);
+        LOGGER.debug("Graph structure: {}", graphStructure);
+        return switch (graphStructure) {
+            case EMPTY -> EmptyVariableReferenceGraph.INSTANCE;
+            case SINGLE_DIRECTIONAL_PARENT -> {
+                var listVariableStateSupply = variableReferenceGraphBuilder.changedVariableNotifier.listVariableStateSupply();
+                if (listVariableStateSupply == null) {
+                    yield buildArbitraryGraph(solutionDescriptor, variableReferenceGraphBuilder, entities, graphCreator);
+                }
+                yield buildSingleDirectionalParentGraph(solutionDescriptor,
+                        variableReferenceGraphBuilder.changedVariableNotifier, listVariableStateSupply,
+                        entities);
+            }
+            case NO_DYNAMIC_EDGES, ARBITRARY ->
+                buildArbitraryGraph(solutionDescriptor, variableReferenceGraphBuilder, entities, graphCreator);
+        };
+    }
+
+    private static <Solution_> VariableReferenceGraph<Solution_> buildSingleDirectionalParentGraph(
```

Comment:
This entire method is very long. Yet it consists of smaller logical pieces that can be extracted to independent subroutines. Please do that - it will make the code easier to understand.